### PR TITLE
Start using UnmanagedCallersOnly in CoreLib

### PR DESCRIPTION
**Pull request by jkotas**

### CODE REVIEW

**Comment by jkotas** [comment]
This is crashing in IL linker with:
```
 Mono.Linker.LinkerFatalErrorException: ILLinker: error IL1005: .Interop.Kernel32.EnumSystemLocalesEx(method,UInt32,Void*,IntPtr): Error processing method '.Interop.Kernel32.EnumSystemLocalesEx(method,UInt32,Void*,IntPtr)' in assembly 'System.Private.CoreLib.dll'
   ---> System.NullReferenceException: Object reference not set to an instance of an object.
     at Mono.Linker.Steps.MarkStep.ProcessInteropMethod(MethodDefinition method)
     at Mono.Linker.Steps.MarkStep.ProcessMethod(MethodDefinition method, DependencyInfo& reason)
     at Mono.Linker.Steps.MarkStep.ProcessQueue()
     --- End of inner exception stack trace ---
     at Mono.Linker.Steps.MarkStep.ProcessQueue()
     at Mono.Linker.Steps.MarkStep.ProcessPrimaryQueue()
     at Mono.Linker.Steps.MarkStep.Process()
     at Mono.Linker.Steps.MarkStep.Process(LinkContext context)
     at Mono.Linker.Pipeline.ProcessStep(LinkContext context, IStep step)
     at Mono.Linker.Pipeline.Process(LinkContext context)
     at Mono.Linker.Driver.Run(ILogger customLogger)
```
I have opened https://github.com/mono/linker/issues/1345
@marek-safar @AaronRobinsonMSFT

**Review by stephentoub — COMMENTED** [review]
Neat.

**Comment by AaronRobinsonMSFT** [comment]
@jkotas Is this something we want to run some benchmarks on? At least so we all have a general idea on the benefit of doing this work or at the very least to get a baseline of what we may expect for wins.

**Comment by jkotas** [comment]
The public entrypoints for what I am touching in this PR are pretty bulky methods. I do not expect to see measurable throughput wins. I expect to see delegate allocation and JITing of the reverse delegate interop interop to disappear with this change. I will check that it is what is happening once it actually builds.

My primary motivation for this change was to validate that the feature actually works end-to-end and shake stress bugs if there are any

**Comment by jkotas** [comment]
@vargaz `UnmanagedCallersOnly` function pointers seems to be broken on Mono. Could you please take look?

Also, PInvokeTable generator has problems with this change, but I should be able to fix that.

**Comment by jkotas** [comment]
I have reverted the Icu change for now to see whether there is anything else that breaks. (The full change is at https://github.com/jkotas/runtime/tree/unmanaged-callers-only-full.)

**Comment by jkotas** [comment]
>  I expect to see delegate allocation and JITing of the reverse delegate interop interop to disappear with this change

I have verified that this does actually happen.

File: tools-local/tasks/mobile.tasks/WasmAppBuilder/PInvokeTableGenerator.cs [hunk 1]
```diff
@@ -187,7 +187,7 @@ void EmitNativeToInterp(StreamWriter w, List<PInvokeCallback> callbacks)
         foreach (var cb in callbacks) {
             var method = cb.Method;
 
-            if (method.ReturnType != typeof(void) && !IsBlittable(method.ReturnType))
+            if (method.ReturnType.Name != "Void" && !IsBlittable(method.ReturnType))
```
Comment (AaronRobinsonMSFT) [inline tools-local/tasks/mobile.tasks/WasmAppBuilder/PInvokeTableGenerator.cs:190]: ~`"Void"`~ -> `typeof(void).Name`?

**Review by AaronRobinsonMSFT — APPROVED** [review]

File: tools-local/tasks/mobile.tasks/WasmAppBuilder/PInvokeTableGenerator.cs (hunk 1 above)
Comment (jkotas) [reply tools-local/tasks/mobile.tasks/WasmAppBuilder/PInvokeTableGenerator.cs:190]: The other places in this file are comparing with `"Void"`, so I went with that for consistency. Ideally, this should really check the full namespace and assembly identity, but it is just a non-shipping helper script so not worth it spending much time on it.

**Comment by jkotas** [comment]
x86 failing with:
```
System.Globalization.Tests.DateTimeFormatInfoLongTimePattern.LongTimePattern_CheckReadingTimeFormatWithSingleQuotes_ICU [SKIP]
      Condition(s) not met: "IsIcuGlobalization"

Assert failure(PID 5328 [0x000014d0], Thread: 4784 [0x12b0]): !pMethod->HasUnmanagedCallersOnlyAttribute()
```

**Comment by vargaz** [comment]
How does UnmanagedCallersOnly fail on mono ? Does it fail on wasm or on desktop ?

**Comment by am11** [comment]
@vargaz, I think it failed on arm64 (desktop). [Here](https://github.com/dotnet/runtime/runs/873936863) are the job runs with ICU interop change. It was [this change](https://github.com/jkotas/runtime/commit/8e6d1ff0e331d802b7d5d6f92b73fffe734bd3ba#diff-fbf6aa30b0c1e6a4afd7688072b4b6df) `delegate* <char*, IntPtr, void> callback`, which is now reverted. [example log](https://helix.dot.net/api/2019-06-17/jobs/1a6d8c1a-509f-442b-a4df-81c974234e06/workitems/JIT.Regression.CLR-x86-JIT.V1-M09-M11/console):
> * Assertion at /__w/1/s/src/mono/mono/metadata/class-init.c:2241, condition `klass->instance_size == instance_size' not met

**Comment by jkotas** [comment]
Yes, this was the failure. You can reproduce it on Linux by checking out https://github.com/jkotas/runtime/tree/unmanaged-callers-only-full and running globalization tests.

**Comment by vargaz** [comment]
Should be fixed by https://github.com/dotnet/runtime/pull/39517
.

**Comment by jkotas** [comment]
Thanks. Does #39517 fixes it on wasm as well?

**Comment by jkotas** [comment]
The new Mono failure pattern is:
```
Unhandled Exception:\nSystem.InvalidProgramException: Invalid IL due to: method System.Globalization.CalendarData:EnumCalendarInfoCallback (char*,intptr) with UnmanagedCallersOnlyAttribute has non-blittable parameters or return type 
```

**Comment by vargaz** [comment]
Can't reproduce that failure.

**Comment by jkotas** [comment]
CoreCLR test failure is:
```
  Starting:    Microsoft.Extensions.Logging.Console.Tests (parallel test collections = on, max threads = 4)

Assert failure(PID 22 [0x00000016], Thread: 40 [0x0028]): m_callCountingInfoByCodeVersionHash.Lookup(codeVersion) == nullptr
    File: /__w/1/s/src/coreclr/src/vm/callcounting.cpp Line: 477
    Image: /root/helix/work/correlation/dotnet
```

**Comment by jkotas** [comment]
Blocked on https://github.com/dotnet/runtime/issues/39604

**Comment by jkotas** [comment]
All failures are due to https://github.com/dotnet/runtime/pull/39082

**Comment by jkotas** [comment]
> Unhandled Exception:\nSystem.InvalidProgramException: Invalid IL due to: method System.Globalization.CalendarData:EnumCalendarInfoCallback (char*,intptr) with UnmanagedCallersOnlyAttribute has non-blittable parameters or return type 

I have extended the Wasm workaround to apply to all Mono targets to work around this. @vargaz You should be able to reproduce this by changing `#if MONO` back `#if TARGETS_WASM`.

**Comment by vargaz** [comment]
Can't reproduce the mono failures.

**Comment by jkotas** [comment]
I have submitted https://github.com/dotnet/runtime/pull/39695 . It should reproduce the failures.

**Pull request merged**